src/operator/contrib/ngraph-inl.h

```diff
@@ -0,0 +1,160 @@
+  ngraph_bridge::NodePtr get_node(const nnvm::Node *n) {
+    if (n) {
+      auto &entry_map = compiler_->get_ngraph().entry_map_;
+      ngraph_bridge::MapEntry tmp{compiler_->get_node_map().at(n).get(), 0};
+      if (entry_map.count(tmp)) {
+        return entry_map[tmp];
+      }
+    }
+    return nullptr;
+  }
+  // is_node_selected queries the ngraph_bridge::Compiler to determine if both
+  // current and next NNVM nodes are supported by nGraph.
+  // This allows us to meld nGraph's analysis with PartitionGraph.
+  bool is_node_selected(const nnvm::Node &n, const nnvm::Node *next = nullptr) {
+    bool selected = false;
+    if (!valid) return selected;
+
+    auto nn = get_node(&n);
+    auto nnext = get_node(next);
+
+    selected = nn && nn->in_ngraph_;
+    if (next) {
+      selected =
+          selected && nnext->in_ngraph_ && nn->subgraph_ == nnext->subgraph_;
+    }
+    return selected;
+  }
+};
+
+class SgNgraphProperty : public SubgraphProperty {
+ public:
+  static SubgraphPropertyPtr Create() {
+    if (ngraph_backend != 0 && ngraph_bridge::ngraph_log_verbose_detail) {
+      LOG(WARNING) << "NGRAPH_BRIDGE: failed to set MXNET_SUBGRAPH_BACKEND"
+                   << std::endl;
+    }
+    return std::make_shared<SgNgraphProperty>();
+  }
+
+  bool NeedGraphAttrs() const override { return true; }
+  // Create a subgraph node based on a symbol
+  nnvm::NodePtr CreateSubgraphNode(
+      const nnvm::Symbol &sym, const int subgraph_id = 0) const override {
+    nnvm::NodePtr n = nnvm::Node::Create();
+    n->attrs.op = Op::Get("_ngraph_subgraph_op");
+    n->attrs.name = "_ngraph_subgraph_op" + std::to_string(subgraph_id);
+    n->attrs.subgraphs.push_back(std::make_shared<nnvm::Symbol>(sym));
+    return n;
+  }
+  // Create a subgraph node based on a graph with inferred shapes, types
+  // and storage types, then compile it with nGraph and store the
+  // ngraph_bridge::Compiler object in NNVM's node attributes for execution.
+  nnvm::NodePtr CreateSubgraphNode(
+      const nnvm::Graph &sg, const int subgraph_id = 0) const override {
+    nnvm::Symbol sym;
+    sym.outputs = sg.outputs;
+    auto n = CreateSubgraphNode(sym, subgraph_id);
+    auto grad_req_map = GetAttr<std::vector<mxnet::OpReqType>>("grad_reqs");
+    auto compiler = std::make_shared<ngraph_bridge::Compiler>(sg, grad_req_map);
+    compiler->GetNgraph();
+    n->attrs.parsed = compiler;
+    return n;
+  }
+  // Create a Subgraph Selector with an embedded ngraph_bridge::Compiler for
+  // nGraph support analysis
+  SubgraphSelectorPtr CreateSubgraphSelector() const override {
+    if (!compiler_) {
+      auto &orig_graph = GetAttr<nnvm::Graph>("graph");
+      auto grad_req_map = GetAttr<std::vector<mxnet::OpReqType>>("grad_reqs");
+      compiler_ = std::make_shared<ngraph_bridge::Compiler>(orig_graph,
+                                                            grad_req_map, true);
+    }
+    return std::make_shared<SgNgraphSelector>(compiler_.get());
```

Comment:
I don't see any reason why we have to pass a raw pointer to SgNgraphSelector. It won't have any noticeable performance improvement.

Reply:
here we use compiler only for query, so we use raw ptr. Right that there is no perf.improvement, we can switch to use shared_ptr.